framework/src/utils/MooseMeshElementConversionUtils.C

```diff
@@ -680,4 +683,546 @@ elementBoundaryInfoCollector(const std::vector<libMesh::BoundaryInfo::BCTuple> &
+                         const bool & external_boundaries_checking)
+{
+  // The base subdomain ID to shift the original elements because of the element type change
+  const auto sid_shift_base = MooseMeshUtils::getNextFreeSubdomainID(mesh);
+
+  // It would be convenient to have a single boundary id instead of a vector.
+  const auto uniform_tmp_bid = MooseMeshUtils::getNextFreeBoundaryID(mesh);
+
+  // Check the boundaries and merge them
+  std::vector<boundary_id_type> boundary_ids;
+  for (const auto & sideset : boundary_names)
+  {
+    if (!MooseMeshUtils::hasBoundaryName(mesh, sideset))
+      throw MooseException("The provided boundary '", sideset, "' was not found within the mesh");
+    boundary_ids.push_back(MooseMeshUtils::getBoundaryID(sideset, mesh));
+    MooseMeshUtils::changeBoundaryId(mesh, boundary_ids.back(), uniform_tmp_bid, false);
+  }
+
+  auto & sideset_map = mesh.get_boundary_info().get_sideset_map();
+  auto side_list = mesh.get_boundary_info().build_side_list();
+
+  std::vector<std::pair<dof_id_type, std::vector<unsigned int>>> elems_list;
+  std::vector<std::set<dof_id_type>> layered_elems_list;
+  layered_elems_list.push_back(std::set<dof_id_type>());
+  // Need to collect the list of elements that need to be converted
+  if (external_boundaries_checking && !mesh.is_prepared())
+    mesh.find_neighbors();
+  for (const auto & side_info : side_list)
+  {
+    if (std::get<2>(side_info) == uniform_tmp_bid)
+    {
+      // Check if the involved side is TRI3 or QUAD4
+      // We do not limit the element type in the input mesh
+      // As long as the involved boundaries only consist of TRI3 and QUAD4 sides,
+      // this generator will work
+      // As the side element of a quadratic element is still a linear element in libMesh,
+      // we need to check the element's default_side_order() first
+      if (mesh.elem_ptr(std::get<0>(side_info))->default_side_order() != 1)
+        throw MooseException(
+            "The provided boundary set contains non-linear side elements, which is not supported.");
+      const auto side_type =
+          mesh.elem_ptr(std::get<0>(side_info))->side_ptr(std::get<1>(side_info))->type();
+      layered_elems_list.back().emplace(std::get<0>(side_info));
+      if (conversion_element_layer_number == 1)
+      {
+        if (side_type == TRI3)
+          continue; // Already TRI3, no need to convert
+        else if (side_type == QUAD4)
+        {
+          if (elems_list.size() && elems_list.back().first == std::get<0>(side_info))
+            elems_list.back().second.push_back(std::get<1>(side_info));
+          else
+            elems_list.push_back(std::make_pair(
+                std::get<0>(side_info), std::vector<unsigned int>({std::get<1>(side_info)})));
+        }
+        else
+          mooseAssert(false,
+                      "Impossible scenario: a linear side element that is neither TRI3 nor QUAD4.");
```

Comment:
well we could have polygon sides of a polyhedron

Reply:
add an error message for polygon side.